### PR TITLE
[Feature] Triton server

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

Support model serving

## Modification

Add triton custom backend
Add demo

### CODE REVIEW

**Comment by irexyc** [comment]
can temporarily use this docker image for testing

`docker pull irexyc/mmmdeploy:triton-22.12`

**Comment by Y-T-G** [comment]
Hey, thanks for this. I wanted to know how do I correctly send multiple bboxes for keypoint-detection inference.

I created a dict for each bbox [here](https://github.com/open-mmlab/mmdeploy/blob/fcdf52f0a504f86e9d3528dd72856a71ef5e28d9/demo/triton/keypoint-detection/grpc_client.py#L84), and used that, but the results are not accurate, although the number of keypoints returned matches the number of bboxes.

**Comment by Y-T-G** [comment]
Also, what does this mean `only support batch dim 1 for single request`? Does this mean the Triton version does not support batch inference?

**Comment by irexyc** [comment]
@Y-T-G 

> I created a dict for each bbox [here](https://github.com/open-mmlab/mmdeploy/blob/fcdf52f0a504f86e9d3528dd72856a71ef5e28d9/demo/triton/keypoint-detection/grpc_client.py#L84) and added to the value list, and used that, but the results are not accurate, although the number of keypoints returned matches the number of bboxes.

Cou you show the visualize result with bboxes? Are the inference result with single bbox looks right?

> Also, what does this mean only support batch dim 1 for single request? Does this mean the Triton version does not support batch inference?

For batch inference of mmdeploy, you can refer to this https://github.com/open-mmlab/mmdeploy/issues/839#issuecomment-1206029364

Triton server support dynamic batcher and sequence batcher. But mmdeploy backend only support dynamic batcher. You can add these lines to config.pbtxt.
```
dynamic_batching {
  max_queue_delay_microseconds: 100
}
```

With [allow_ragged_batch](https://github.com/open-mmlab/mmdeploy/blob/fcdf52f0a504f86e9d3528dd72856a71ef5e28d9/demo/triton/keypoint-detection/serving/model/config.pbtxt#L8) and `dynamic_batching`, mmdeploy backend can receive a batch of requests for each inference step (therefore, you don't have to construct normal batch input like b x c x h x w. you only need send c x h x w to triton server and let it to collect batch requests.) 

In summary, to use mmdeploy triton backend with batch inference, you have to:
1) convert the model with batch inference support and edit the pipeline.json
2) add `dynamic_batching` to config.pbtxt

**Comment by Y-T-G** [comment]
@irexyc

> Cou you show the visualize result with bboxes? Are the inference result with single bbox looks right?

Yes with single bbox the inference is correct. But if I add more than one bbox, the outputs don't make any sense.

I only visualize the nose, left wrist and right wrist keypoints.

This is how it looks when I add more than 1:
image

This is how it looks like when I do individually, cropping each bbox and sending each for inference separately:
image

The input for multiple `bbox` looks like this:

```
{
   "type":"PoseBbox",
   "value":[
      {
         "bbox":[
            866,
            47,
            896,
            101
         ]
      },
      {
         "bbox":[
            48,
            65,
            73,
            125
         ]
      },
      {
         "bbox":[
            425,
            32,
            447,
            97
         ]
      },
      {
         "bbox":[
            400,
            36,
            422,
            98
         ]
      },
      {
         "bbox":[
            193,
            262,
            230,
            334
         ]
      },
      {
         "bbox":[
            602,
            57,
            627,
            126
         ]
      },
      {
         "bbox":[
            950,
            53,
            959,
            101
         ]
      },
      {
         "bbox":[
            571,
            239,
            607,
            313
         ]
      },
      {
         "bbox":[
            585,
            66,
            611,
            130
         ]
      },
      {
         "bbox":[
            228,
            254,
            258,
            322
         ]
      },
      {
         "bbox":[
            116,
            14,
            146,
            73
         ]
      },
      {
         "bbox":[
            547,
            28,
            569,
            81
         ]
      },
      {
         "bbox":[
            899,
            80,
            930,
            142
         ]
      },
      {
         "bbox":[
            173,
            246,
            202,
            320
         ]
      },
      {
         "bbox":[
            463,
            95,
            493,
            161
         ]
      },
      {
         "bbox":[
            356,
            111,
            385,
            184
         ]
      },
      {
         "bbox":[
            379,
            211,
            410,
            293
         ]
      },
      {
         "bbox":[
            286,
            145,
            317,
            213
         ]
      },
      {
         "bbox":[
            925,
            325,
            960,
            420
         ]
      },
      {
         "bbox":[
            739,
            250,
            782,
            339
         ]
      },
      {
         "bbox":[
            338,
            301,
            378,
            398
         ]
      },
      {
         "bbox":[
            732,
            175,
            767,
            246
         ]
      },
      {
         "bbox":[
            123,
            235,
            165,
            321
         ]
      },
      {
         "bbox":[
            695,
            355,
            746,
            445
         ]
      },
      {
         "bbox":[
            568,
            443,
            618,
            535
         ]
      }
   ]
}
```

**Comment by Y-T-G** [comment]
> For batch inference of mmdeploy, you can refer to this [#839 (comment)](https://github.com/open-mmlab/mmdeploy/issues/839#issuecomment-1206029364)
> 
> Triton server support dynamic batcher and sequence batcher. But mmdeploy backend only support dynamic batcher. You can add these lines to config.pbtxt.
> 
> ```
> dynamic_batching {
>   max_queue_delay_microseconds: 100
> }
> ```
> 
> With [allow_ragged_batch](https://github.com/open-mmlab/mmdeploy/blob/fcdf52f0a504f86e9d3528dd72856a71ef5e28d9/demo/triton/keypoint-detection/serving/model/config.pbtxt#L8) and `dynamic_batching`, mmdeploy backend can receive a batch of requests for each inference step (therefore, you don't have to construct normal batch input like b x c x h x w. you only need send c x h x w to triton server and let it to collect batch requests.)
> 
> In summary, to use mmdeploy triton backend with batch inference, you have to:
> 
> 1. convert the model with batch inference support and edit the pipeline.json
> 2. add `dynamic_batching` to config.pbtxt

I am not sure if this works. I don't see any improvements when I do this after checking with `perf_analyzer` for ResNet18:
```
Inferences/Second vs. Client Average Batch Latency
Concurrency: 2, throughput: 41.6086 infer/sec, latency 47983 usec
Concurrency: 3, throughput: 41.8305 infer/sec, latency 71626 usec
Concurrency: 4, throughput: 41.775 infer/sec, latency 95672 usec
Concurrency: 5, throughput: 41.2752 infer/sec, latency 120931 usec
Concurrency: 6, throughput: 41.7747 infer/sec, latency 143440 usec
Concurrency: 7, throughput: 41.7748 infer/sec, latency 167467 usec
Concurrency: 8, throughput: 41.6641 infer/sec, latency 191807 usec
```
It supports batching in the `json`.

I can see better improvements by launching multiple model instances using:
```
instance_group [ 
  { 
    count: 4
    kind: KIND_GPU 
  }
]
```
Inferences/Second vs. Client Average Batch Latency
Concurrency: 2, throughput: 62.7163 infer/sec, latency 31838 usec
Concurrency: 3, throughput: 65.1612 infer/sec, latency 46015 usec
Concurrency: 4, throughput: 79.328 infer/sec, latency 50415 usec
Concurrency: 5, throughput: 84.3826 infer/sec, latency 59160 usec
Concurrency: 6, throughput: 90.2152 infer/sec, latency 66516 usec
Concurrency: 7, throughput: 89.4926 infer/sec, latency 78322 usec
Concurrency: 8, throughput: 88.104 infer/sec, latency 90731 usec
```

I think dynamic_batcher depends on sequence_batching. But since each request is handled separately in `instance_state.cpp`, dynamic_batching will not have any effect. To have an effect, the requests have to batched and then inferred all at once.
